internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,195 @@
+	}
+	return r, nil
+}
+
+func targetURIResolver(target resolver.Target, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder, resolver *delegatingResolver) (resolver.Resolver, error) {
+	targetBuilder := targetResolverBuilder
+	if targetBuilder == nil {
+		return nil, fmt.Errorf("resolver for target scheme %q not found", target.URL.Scheme)
+	}
+	return targetBuilder.Build(target, &innerClientConn{resolver, "target"}, opts)
+}
+
+func proxyURIResolver(proxyURL *url.URL, opts resolver.BuildOptions, presolver *delegatingResolver) (resolver.Resolver, error) {
+	scheme := "dns"
+	proxyBuilder := resolver.Get(scheme)
+	if proxyBuilder == nil {
+		return nil, fmt.Errorf("resolver for proxy not found")
+	}
+	host := "dns:///" + proxyURL.Host
+	u, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+
+	proxyTarget := resolver.Target{URL: *u}
+	return proxyBuilder.Build(proxyTarget, &innerClientConn{presolver, "proxy"}, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.ProxyResolver != nil {
+		r.ProxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+	}
+	if r.ProxyResolver != nil {
+		r.ProxyResolver.Close()
+	}
+}
+
+// UpdateState intercepts state updates from the childtarget and proxy resolvers.
+func (icc *innerClientConn) UpdateState(state resolver.State) error {
+	icc.mu.Lock()
+	defer icc.mu.Unlock()
+
+	var curState resolver.State
+	if icc.resolverType == "target" {
+		icc.targetAddrs = state.Addresses
+		curState = state
+	}
+	if icc.resolverType == "proxy" {
+		icc.proxyAddrs = state.Addresses
+	}
+
+	if len(icc.targetAddrs) == 0 || len(icc.proxyAddrs) == 0 {
+		return nil
+	}
+
+	var addresses []resolver.Address
+	for _, proxyAddr := range icc.proxyAddrs {
+		for _, targetAddr := range icc.targetAddrs {
+			attr := attributes.New("proxyConnectAddr", targetAddr.Addr)
+			if icc.proxyURL.User != nil {
+				attr = attr.WithValue("user", icc.proxyURL.User)
+			}
+			addresses = append(addresses, resolver.Address{
+				Addr:       proxyAddr.Addr,
+				Attributes: attr,
+			})
+		}
+	}
+	// Set the addresses in the current state.
+	curState.Addresses = addresses
+	return icc.cc.UpdateState(curState)
+}
+
+// ReportError intercepts errors from the child DNS resolver.
+func (icc *innerClientConn) ReportError(err error) {
+	icc.cc.ReportError(err)
+}
+
+func (icc *innerClientConn) NewAddress(addrs []resolver.Address) {
+}
+
+// ParseServiceConfig parses the provided service config and returns an
+// object that provides the parsed config.
+func (icc *innerClientConn) ParseServiceConfig(serviceConfigJSON string) *serviceconfig.ParseResult {
+	fmt.Printf("config received by delegating resolver : %v \n", serviceConfigJSON)
+	if icc.resolverType == "target" {
+		return icc.cc.ParseServiceConfig(serviceConfigJSON)
+	}
+	return nil
```

Comment:
Hmm... Should this be delegated to the `icc.cc` as well? If not, can you please explain why? I haven't thought through this enough. So, i'm relying on you to provide the correct answer.

Reply:
Yes I think it should be delegated to icc.cc as it a necessary method and the target resolver could call the method , and when that happens, the method in delegating resolver will be called which we should pass to icc.cc to get the correct parsed config.